### PR TITLE
RT: train cancelling

### PR DESCRIPTION
**Pull request by antoine-de**
handle realtime message to cancel a train (VehicleJourney)

To query on realtime data a new `journeys` parameter has been added:
`data_freshess` (the name can be discussed) that can take 3 values:
- base_schedule 
- adapted_schedule
- realtime

The `disruption_active` parameter is thus now deprecated

### CODE REVIEW

File: source/kraken/realtime.cpp [hunk 1]
```diff
@@ -29,11 +29,53 @@ www.navitia.io
...
+    if (! vj) {
+        LOG4CPLUS_INFO(log, "unknown vehicle journey " << trip.trip_id());
+        // TODO for trip().ADDED, we'll need to create a new VJ
+        return;
+    }
+
+    auto circulation_date = boost::gregorian::from_undelimited_string(trip.start_date());
+
+    if (trip.CANCELED) {
+        cancel_vj(vj, circulation_date, trip_update, data);
```
Comment (TeXitoi) [inline source/kraken/realtime.cpp]: I think you need to loop on every vj of the meta vj.

File: source/jormungandr/jormungandr/interfaces/v1/Journeys.py [hunk 2]
```diff
@@ -624,6 +626,10 @@ def get(self, region=None, lon=None, lat=None, uri=None):
             args['max_bss_duration_to_pt'] = args['max_duration_to_pt']
             args['max_car_duration_to_pt'] = args['max_duration_to_pt']
 
+        if args['data_freshness'] is None:
+            # retrocompatibilty handling
+            args['data_freshness'] = 'adapted' if args['disruption_active'] is True else 'theoric'
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/Journeys.py]: it should be `theoretical`.. let's keep it since it's used everywhere LOL

File: source/routing/dataraptor.cpp [hunk 3]
```diff
@@ -90,29 +90,20 @@ void dataRAPTOR::load(const type::PT_Data &data)
...
-                return true;
-            });
+    for (auto level_cont: jp_validity_patterns) {
+        const auto rt_level = level_cont.first;
+        auto& jp_vp = level_cont.second;
+        jp_vp.assign(366, boost::dynamic_bitset<>(data.journey_patterns.size()));
+        for(const type::JourneyPattern* journey_pattern : data.journey_patterns) {
+            for(int i=0; i<=365; ++i) {
+                journey_pattern->for_each_vehicle_journey([&](const nt::VehicleJourney& vj) {
+                    if(vj.validity_patterns[rt_level]->check2(i)) {
```
Comment (TeXitoi) [inline source/routing/dataraptor.cpp]: spacingsss

File: source/routing/raptor.cpp [hunk 4]
```diff
@@ -412,16 +412,8 @@ void RAPTOR::set_valid_jp_and_jpp(
...
-    case nt::RTLevel::Theoric:
-        valid_journey_patterns = data.dataRaptor->jp_validity_patterns[date];
-        break;
-    case nt::RTLevel::Adapted:
-        valid_journey_patterns = data.dataRaptor->jp_adapted_validity_pattern[date];
-        break;
-    case nt::RTLevel::RealTime:
-        throw navitia::exception("rt not implemented yet");
-    }
+
```
Comment (TeXitoi) [inline source/routing/raptor.cpp]: empty line

File: source/jormungandr/jormungandr/scenarios/default.py [hunk 5]
```diff
@@ -100,10 +100,14 @@ def parse_journey_request(self, requested_type, request):
         req.journeys.max_duration = request["max_duration"]
         req.journeys.max_transfers = request["max_transfers"]
         req.journeys.wheelchair = request["wheelchair"]
-        if request["disruption_active"]:
+
+        if request["data_freshness"] == 'realtime':
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/default.py]: single quote

File: source/jormungandr/tests/kirin_realtime_tests.py [hunk 6]
```diff
@@ -45,21 +45,56 @@ def _make_mock_item(self, *args, **kwargs):
...
+        response = self.query_region(journey_basic_query + "&data_freshness=realtime")
 
-        response = self.query_region('vehicle_journeys')
-        assert len(response) > 0
+        # with no cancellation, we have 2 jounrneys, one direct and one with the vj:A:0
+        assert _get_arrivals(response) == ['20120614T080222', '20120614T080435']
+        assert _get_used_vj(response) == [['vj:A:0'], []]
 
-        #TODO
+        self.send_mock("vj:A:0", "20120614", 'cancelled')
```
Comment (xlqian) [inline source/jormungandr/tests/kirin_realtime_tests.py]: Minor: cancelled or canceled?

File: source/jormungandr/tests/kirin_realtime_tests.py [hunk 7]
```diff
@@ -70,13 +105,13 @@ def make_mock_kirin_item(*args, **kwargs):
...
-
-    stop_time = trip_update.stop_time_update.add()
-    stop_time.stop_id = "stop_point:OCE:SP:TrainTER-87212027"
-    stop_time.arrival.time = utils.str_to_time_stamp("20150728T1721")
-    stop_time.departure.time = utils.str_to_time_stamp("20150728T1721")
+    trip.trip_id = vj_id  #
+    trip.start_date = date
+
+    if status == 'canceled':
+        trip.schedule_relationship = gtfs_realtime_pb2.TripDescriptor.CANCELED
```
Comment (xlqian) [inline source/jormungandr/tests/kirin_realtime_tests.py:113]: Minor: cancelled or canceled?

File: source/kraken/realtime.cpp [hunk 8]
```diff
@@ -29,11 +29,53 @@ www.navitia.io
...
+    auto vj = find_or_default(trip.trip_id(), data.pt_data->vehicle_journeys_map);
+    if (! vj) {
+        LOG4CPLUS_INFO(log, "unknown vehicle journey " << trip.trip_id());
+        // TODO for trip().ADDED, we'll need to create a new VJ
+        return;
+    }
+
+    auto circulation_date = boost::gregorian::from_undelimited_string(trip.start_date());
+
+    if (trip.CANCELED) {
```
Comment (xlqian) [inline source/kraken/realtime.cpp:80]: `trip.CANCELED` has been tested, no?

File: source/kraken/tests/realtime_test.cpp [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+#include "type/gtfs-realtime.pb.h"
+#include "type/pt_data.h"
+#include "kraken/realtime.h"
+#include "ed/build_helper.h"
+#include "tests/utils_test.h"
+#include "routing/raptor.h"
+
+namespace nt = navitia::type;
+
+transit_realtime::TripUpdate make_cancellation_message(std::string vj_uri, std::string date) {
```
Comment (xlqian) [inline source/kraken/tests/realtime_test.cpp]: `const std::string&`

File: source/kraken/realtime.cpp (hunk 8 above)
Comment (antoine-de) [inline source/kraken/realtime.cpp:80]: yes, but the first one is temporary, since for the moment we handle only the CANCELED tag

File: source/kraken/tests/realtime_test.cpp (hunk 9 above)
Comment (antoine-de) [inline source/kraken/tests/realtime_test.cpp]: I don't think it's worth it since it is always used with a lvalue

File: source/jormungandr/tests/kirin_realtime_tests.py (hunk 7 above)
Comment (antoine-de) [inline source/jormungandr/tests/kirin_realtime_tests.py:113]: http://www.quickanddirtytips.com/education/grammar/canceled-or-cancelled :)

File: source/kraken/tests/realtime_test.cpp (hunk 9 above)
Comment (TeXitoi) [inline source/kraken/tests/realtime_test.cpp]: It's not worst to use const&, and it's better in the general case.

You also need to put this method static.

File: source/kraken/tests/realtime_test.cpp [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+    return trip_update;
+}
+
+BOOST_AUTO_TEST_CASE(simple_train_cancellation) {
+    ed::builder b("20150928");
+    b.vj("A", "000001", "", true, "vj:1")("stop1", "08:00"_t)("stop2", "09:00"_t);
+
+    transit_realtime::TripUpdate trip_update = make_cancellation_message("vj:1", "20150928");
+    auto& pt_data = b.data->pt_data;
```
Comment (xlqian) [inline source/kraken/tests/realtime_test.cpp]: auto* pt_data

File: source/kraken/realtime.cpp (hunk 8 above)
Comment (xlqian) [inline source/kraken/realtime.cpp:80]: ok

File: source/jormungandr/jormungandr/interfaces/v1/Journeys.py (hunk 2 above)
Comment (pbougue) [inline source/jormungandr/jormungandr/interfaces/v1/Journeys.py]: As it is the first time we expose it I vote for a change.
Other suggestion from Pascal : `revised` instead of adapted

File: source/routing/dataraptor.cpp [hunk 11]
```diff
@@ -90,29 +90,20 @@ void dataRAPTOR::load(const type::PT_Data &data)
...
-                    jp_adapted_validity_pattern[i].set(journey_pattern->idx);
-                    return false;
-                }
-                return true;
-            });
+    for (auto level_cont: jp_validity_patterns) {
+        const auto rt_level = level_cont.first;
+        auto& jp_vp = level_cont.second;
+        jp_vp.assign(366, boost::dynamic_bitset<>(data.journey_patterns.size()));
+        for(const type::JourneyPattern* journey_pattern : data.journey_patterns) {
```
Comment (xlqian) [inline source/routing/dataraptor.cpp]: const auto*? for consistence?

File: source/jormungandr/tests/kirin_realtime_tests.py (hunk 7 above)
Comment (xlqian) [inline source/jormungandr/tests/kirin_realtime_tests.py:113]: It was a question more about consistence than english grammar :)

File: source/ed/ed_reader.cpp [hunk 12]
```diff
@@ -822,6 +822,7 @@ void EdReader::fill_vehicle_journeys(nt::Data& data, pqxx::work& work){
         if(!const_it["validity_pattern_id"].is_null()){
             vj->validity_patterns[RTLevel::Theoric] =
                     validity_pattern_map[const_it["validity_pattern_id"].as<idx_t>()];
+            vj->validity_patterns[RTLevel::RealTime] = vj->validity_patterns[RTLevel::Theoric];
```
Comment (xlqian) [inline source/ed/ed_reader.cpp]: so, no real time information in db just now or there will never be?

File: source/kraken/tests/realtime_test.cpp [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    auto compute = [&](nt::RTLevel level) {
+        return raptor.compute(pt_data->stop_areas_map.at("stop1"), pt_data->stop_areas_map.at("stop2"),
+                              "08:00"_t, 0, navitia::DateTimeUtils::inf, level, true);
+    };
+
+    //on the theoric level, we should get one solution
+    auto res = compute(nt::RTLevel::Theoric);
+    BOOST_REQUIRE_EQUAL(res.size(), 1);
+
+    //on the realtime level, we should also gt one solution, since for the moment there is no cancellation
```
Comment (pbougue) [inline source/kraken/tests/realtime_test.cpp]: gt > `get`

File: source/ed/ed_reader.cpp (hunk 12 above)
Comment (antoine-de) [inline source/ed/ed_reader.cpp]: maybe there will, maybe there won't :)
why ?

Comment (pbougue) [inline source/ed/ed_reader.cpp]: At minimum, add a little comment for that.
But all this consistency-and-C++-model should be in a special builder (or in meta-vj), not in ED, builder, chaos and Kirin readers anymore

Comment (antoine-de) [inline source/ed/ed_reader.cpp]: I think this should be done in another PR (at least to wait for the JourneyPattern refactoring).

Link: http://jira.canaltp.fr/browse/NAVP-299

File: source/jormungandr/tests/kirin_realtime_tests.py (hunk 7 above)
Comment (antoine-de) [inline source/jormungandr/tests/kirin_realtime_tests.py:113]: yes you're right, I choose the english version (canceled)

File: source/jormungandr/jormungandr/interfaces/v1/Journeys.py (hunk 2 above)
Comment (antoine-de) [inline source/jormungandr/jormungandr/interfaces/v1/Journeys.py]: I personally vote for `scheduled` and `amended`, but I think it's better to wait for andrew

File: source/kraken/realtime.cpp (hunk 8 above)
Comment (TeXitoi) [inline source/kraken/realtime.cpp]: So you want a switch here...

**Comment by TeXitoi** [comment]
OK for me.
